File: saksflyt/src/main/java/no/nav/melosys/saksflyt/impl/ProsessinstansFerdigListener.java
package no.nav.melosys.saksflyt.impl;

import java.util.Comparator;
import java.util.Set;

import no.nav.melosys.domain.saksflyt.ProsessStatus;
import no.nav.melosys.domain.saksflyt.Prosessinstans;
import no.nav.melosys.domain.saksflyt.SedLåsReferanse;
import no.nav.melosys.repository.ProsessinstansRepository;
import no.nav.melosys.saksflyt.api.ProsessinstansBehandler;
import no.nav.melosys.service.saksflyt.ProsessinstansFerdigEvent;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.context.event.EventListener;
import org.springframework.stereotype.Component;

@Component
public class ProsessinstansFerdigListener {

    private static final Logger log = LoggerFactory.getLogger(ProsessinstansFerdigListener.class);

    private final ProsessinstansRepository prosessinstansRepository;
    private final ProsessinstansBehandler prosessinstansBehandler;

    public ProsessinstansFerdigListener(ProsessinstansRepository prosessinstansRepository,
                                        ProsessinstansBehandler prosessinstansBehandler) {
        this.prosessinstansRepository = prosessinstansRepository;
        this.prosessinstansBehandler = prosessinstansBehandler;
    }

    @EventListener
    public void prosessinstansFerdig(ProsessinstansFerdigEvent prosessinstansFerdigEvent) {
        log.info("Prosessinstans {} ferdig", prosessinstansFerdigEvent.getUuid());
        if (prosessinstansFerdigEvent.getLåsReferanse() != null && !finnesAktivReferanse(prosessinstansFerdigEvent.getLåsReferanse())) {
            startNesteProsessinstans(prosessinstansFerdigEvent);
        }
    }

    private boolean finnesAktivReferanse(String referanse) {
        return prosessinstansRepository.existsByStatusNotInAndLåsReferanse(Set.of(ProsessStatus.FERDIG), referanse);
    }

    private void startNesteProsessinstans(ProsessinstansFerdigEvent prosessinstansFerdigEvent) {
        log.info("Forsøker å starte neste prosessinstans, låsreferanse {}", prosessinstansFerdigEvent.getLåsReferanse());
        var ferdigReferanse = new SedLåsReferanse(prosessinstansFerdigEvent.getLåsReferanse());

        var prosessinstanserPåVent = prosessinstansRepository.findAllByStatus(ProsessStatus.PÅ_VENT);

        prosessinstanserPåVent.stream()
            .filter(p -> harSammeReferanse(p, ferdigReferanse))
            .min(Comparator.comparing(Prosessinstans::getRegistrertDato))
            .ifPresent(this::oppdaterStatusOgBehandleProsessinstans);
    }

    private void oppdaterStatusOgBehandleProsessinstans(Prosessinstans prosessinstans) {
        log.info("Prosessinstans {} startes opp etter å ha vært på vent", prosessinstans.getId());
        prosessinstans.setStatus(ProsessStatus.KLAR);
        prosessinstansRepository.save(prosessinstans);
        prosessinstansBehandler.behandleProsessinstans(prosessinstans);
    }

    private boolean harSammeReferanse(Prosessinstans prosessinstans, SedLåsReferanse ferdigLåsreferanse) {
        var låsReferanse = new SedLåsReferanse(prosessinstans.getLåsReferanse());
        return låsReferanse.getReferanse().equals(ferdigLåsreferanse.getReferanse());
    }
}

File: saksflyt/src/main/java/no/nav/melosys/saksflyt/ProsessinstansFerdigListener.kt
package no.nav.melosys.saksflyt

import mu.KotlinLogging
import no.nav.melosys.saksflytapi.domain.ProsessStatus
import no.nav.melosys.saksflytapi.domain.Prosessinstans
import no.nav.melosys.saksflytapi.domain.SedLåsReferanse
import org.springframework.context.event.EventListener
import org.springframework.stereotype.Component
import java.time.LocalDateTime
import java.util.Set

private val log = KotlinLogging.logger { }
@Component
class ProsessinstansFerdigListener(
    private val prosessinstansRepository: ProsessinstansRepository,
    private val prosessinstansBehandler: ProsessinstansBehandler
) {
    @EventListener
    fun prosessinstansFerdig(prosessinstansFerdigEvent: ProsessinstansFerdigEvent) {
        log.info("Prosessinstans {} ferdig", prosessinstansFerdigEvent.uuid)
        if (prosessinstansFerdigEvent.låsReferanse != null && !finnesAktivReferanse(prosessinstansFerdigEvent.låsReferanse)) {
            startNesteProsessinstans(prosessinstansFerdigEvent)
        }
    }

    private fun finnesAktivReferanse(referanse: String): Boolean {
        return prosessinstansRepository.existsByStatusNotInAndLåsReferanse(Set.of(ProsessStatus.FERDIG), referanse)
    }

    private fun startNesteProsessinstans(prosessinstansFerdigEvent: ProsessinstansFerdigEvent) {
        log.info("Forsøker å starte neste prosessinstans, låsreferanse {}", prosessinstansFerdigEvent.låsReferanse)
        val ferdigReferanse = SedLåsReferanse(prosessinstansFerdigEvent.låsReferanse)

        val prosessinstanserPåVent = prosessinstansRepository.findAllByStatus(ProsessStatus.PÅ_VENT)

        prosessinstanserPåVent.stream()
            .filter { p: Prosessinstans -> harSammeReferanse(p, ferdigReferanse) }
            .min(Comparator.comparing { obj: Prosessinstans -> obj.registrertDato })
            .ifPresent { prosessinstans: Prosessinstans -> this.oppdaterStatusOgBehandleProsessinstans(prosessinstans) }
    }

    private fun oppdaterStatusOgBehandleProsessinstans(prosessinstans: Prosessinstans) {
        log.info("Prosessinstans {} startes opp etter å ha vært på vent", prosessinstans.id)
        prosessinstans.status = ProsessStatus.KLAR
        prosessinstans.endretDato = LocalDateTime.now()
        prosessinstansRepository.save(prosessinstans)
        prosessinstansBehandler.behandleProsessinstans(prosessinstans)
    }

    private fun harSammeReferanse(prosessinstans: Prosessinstans, ferdigLåsreferanse: SedLåsReferanse): Boolean {
        val låsReferanse = SedLåsReferanse(prosessinstans.låsReferanse)
        return låsReferanse.referanse == ferdigLåsreferanse.referanse
    }
}
